microsoft-edge/progressive-web-apps-chromium/how-to/widgets.md

```diff
@@ -0,0 +1,495 @@
+
+
+<!-- ====================================================================== -->
+<!-- widgets are too early in the making for MDN. But there is enough complexity with the APIs that a reference doc is needed.
+For now we are documenting all functions and objects in this section. When widgets become standard, move this to MDN and remove the section. -->
+## Service Worker API reference
+
+The service worker global object (or [ServiceWorkerGlobalScope](https://developer.mozilla.org/docs/Web/API/ServiceWorkerGlobalScope)) contains a `widgets` attribute that exposes the following Promise-based methods:
+
+| Method | Description | Parameters | Return value |
+|:--- |:--- |:--- |:--- |
+| `getByTag(tag)` | Get a widget by tag | The widget tag | A Promise that resolves to the [widget object](#widget-object) that matches the tag, or `undefined`. |
+| `getByInstanceId(id)` | Get a widget by instance ID | The widget instance ID | A Promise that resolves to the corresponding [widget object](#widget-object), or undefined.
+| `getByHostId(id)` | Get widgets by host ID  | The host ID | An array of [widget objects](#widget-object) found in that host.
+| `matchAll(options)` | Get widgets by matching options | A [widgetOptions object](#widgetoptions-object) | A Promise that resolves to an array of [widget objects](#widget-object) that match the `options` criteria. |
+| `updateByInstanceId(id, payload)` | Update a widget by instance ID  | The instance ID, and a [widgetPayload object](#widgetpayload-object) | A Promise that resolves to `undefined` or `Error`.
+| `updateByTag(tag, payload)` | Update a widget by tag | The widget tag, and a [widgetPayload object](#widgetpayload-object) | A Promise that resolves to undefined or Error.
+
+The service worker global object also defines the `widgetclick` event that's fired when the widget is interacted with. For more information, see the [widgetEvent object](#widgetevent-object) definition below.
+
+#### widget object
+
+Each widget is represented as a `widget` object, which contains the following properties:
+
+* `installable`: A Boolean indicating wither the widget is installable.
+* `definition`: A [widgetDefinition object](#widgetdefinition-object).
+* `instances`: An array of [widgetInstance objects](#widgetinstance-object) representing the current state of each instance of the widget.
+
+#### widgetOptions object
+
+When using `matchAll(options)` to get multiple widgets, a `widgetOptions` object is necessary to filter which widgets to return. The `widgetOptions` object contains the following properties, all of which are optional:
+
+* `installable`: A Boolean that indicates if the widget is considered installable.
+* `installed`: A Boolean that indicates if the widget has been installed in the widget host.
+* `tag`: The tag of the widget.
```

Comment:
Some fields above have types specified - boolean, array, etc. Should the tag here and some other ones have their types listed too?